### PR TITLE
[primTorch fix] use _maybe_convert_to_dtype

### LINKED ISSUE #84561: Unnecessary uses of `copy_to` in refs implementations
### 🐛 Describe the bug

Many operations, for example `linalg.vector_norm`, use `x = prims.convert_element_type(x, computation_dtype)`. This unconditionally calls to `copy_to`. We should instead use `x = _maybe_convert_to_dtype(x, computation_dtype)`

### Versions

master

cc @ezyang @mruberry @ngimel

### PR DESCRIPTION
**Pull request by khushi-411**
Fixes #84561

- [x] fix lint tests

cc: @Lezcano!!

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
LGTM!

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -1280,8 +1280,8 @@ def isclose(
     # If the values of the integer tensors cannot be exactly represented
     # by the default scalar type then this may cause an incorrect result.
     if not utils.is_float_dtype(a.dtype) and not utils.is_complex_dtype(a.dtype):
-        a = prims.convert_element_type(a, torch.get_default_dtype())
-        b = prims.convert_element_type(b, torch.get_default_dtype())
+        a = _maybe_convert_to_dtype(a, torch.get_default_dtype())
```
Comment (mruberry) [inline torch/_refs/__init__.py:1283]: In this case the conversion will always occur since the default dtype is always a floating point type

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -1298,14 +1298,14 @@ def _lcm(a: TensorLikeType, b: TensorLikeType):
     dtype = a.dtype
     promote_to_int = dtype in (torch.int8, torch.int16)
     if promote_to_int:
-        a = prims.convert_element_type(a, torch.int32)
-        b = prims.convert_element_type(b, torch.int32)
+        a = _maybe_convert_to_dtype(a, torch.int32)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1301]: In this case the promotion will definitely occur because we know `a` has a different dtype than `torch.int32`

Comment (mruberry) [reply torch/_refs/__init__.py:1301]: Is this implementation correct thought? It seems odd that it's only checking the dtype of `a` -- what about the dtype of `b`? Also it could use more comments.

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -1771,7 +1771,7 @@ def _reduction(
         return _safe_copy_out(copy_from=result, copy_to=out)  # type: ignore[arg-type]
 
     if result.dtype != result_dtype and result_dtype is not None:
-        result = prims.convert_element_type(result, result_dtype)
+        result = _maybe_convert_to_dtype(result, result_dtype)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1774]: This conversion is known to occur because the dtypes are compared in the previous line

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -1798,7 +1798,7 @@ def all(
 
     # Preserves uint8 -- probably a legacy mask thing
     if a.dtype is torch.uint8:
-        return prims.convert_element_type(result, torch.uint8)
+        return _maybe_convert_to_dtype(result, torch.uint8)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1801]: I think result is known to not be uint8 in this case

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -1819,7 +1819,7 @@ def any(
...
```
Comment (mruberry) [inline torch/_refs/__init__.py:1822]: I think this conversion will always occur

File: torch/_refs/fft.py [hunk 6]
```diff
@@ -78,7 +81,7 @@ def _maybe_promote_tensor_fft(
     new_type = _promote_type_fft(cur_type, require_complex)
     if cur_type == new_type:
         return t
-    return prims.convert_element_type(t, new_type)
+    return _maybe_convert_to_dtype(t, new_type)
```
Comment (mruberry) [inline torch/_refs/fft.py:84]: This conversion is known to occur -- the previous check handles the case where they're the same. `_maybe_convert_to_dtype` can be used here but it should replace the `if` above

**Review by mruberry — COMMENTED** [review]
Hey @khushi-411! Scrubbing for unnecessary conversions is great, but these are going to need another close look. I think a lot of these conversions are OK to always occur due to the logic around them. I'm also a little confused by what's going on with `_lcm` -- that may need investigation

File: torch/_refs/__init__.py (hunk 1 above)
Comment (khushi-411) [reply torch/_refs/__init__.py]: Hi, @mruberry! I checked into the functions your point completely makes sense to me. Thanks for pointing it out :)

File: torch/_refs/__init__.py (hunk 2 above)
Comment (khushi-411) [reply torch/_refs/__init__.py]: Yes, I think so after reading this (originally investigated in https://github.com/pytorch/pytorch/pull/82950#issue-1331192753).

> lcm is meant to use integer maths, but the use of true_divide
introduces a promotion to float and thus a loss of precision.
This also introduces promoting low precision integers to int32 which
is required for 100% consistency with the C++ implementation.

Thanks!

File: torch/_refs/__init__.py (hunk 3 above)
Comment (khushi-411) [reply torch/_refs/__init__.py]: Thanks for the explanation! Got that.

File: torch/_refs/fft.py (hunk 6 above)
Comment (khushi-411) [reply torch/_refs/fft.py:79]: Agreed with you :) Updated it. Thanks

**Comment by khushi-411** [comment]
Hi, @mruberry! Gentle ping :) Please take another look.

File: torch/_refs/__init__.py [hunk 7]
```diff
@@ -1360,7 +1362,7 @@ def _lcm(a: TensorLikeType, b: TensorLikeType):
     # Avoid division by zero in case gcd(0, 0) == 0
     g = torch.where(g == 0, 1, g)
     res = torch.abs(prims.div(a, g) * b)
-    return res if not promote_to_int else prims.convert_element_type(res, dtype)
+    return res if not promote_to_int else _maybe_convert_to_dtype(res, dtype)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1365]: I think this conversion is unconditional because it's guarded on `promote_to_int`

**Review by mruberry — APPROVED** [review]
Cool! See one inline note for review

File: torch/_refs/__init__.py (hunk 7 above)
Comment (khushi-411) [reply torch/_refs/__init__.py:1365]: Undone it

**Comment by khushi-411** [comment]
Thank you very much for reviewing, @lezcano @mruberry!!
